packages/streamid/src/event-id.ts

```diff
@@ -0,0 +1,202 @@
+import { CID } from 'multiformats/cid'
+import { base36 } from 'multiformats/bases/base36'
+import { hash as sha256 } from '@stablelib/sha256';
+import varint from 'varint'
+import * as cbor from 'cborg'
+import * as u8a from 'uint8arrays'
+import { STREAMID_CODEC } from './constants.js'
+import { Memoize } from 'mapmoize'
+
+const TAG = Symbol.for('@ceramicnetwork/streamid/EventID')
+
+const DAG_CBOR_CODEC = 0x30
+const STREAMID_CODEC_LEN = 2
+const DAG_CBOR_CODEC_LEN = 1
+
+const NETWORK: Record<string, number> = {
+  "mainnet":	0x00,
+  "testnet-clay":	0x01,
+  "dev-unstable":	0x02,
+  "inmemory":	0xFF
+}
+
+function networkByName(net: string):number {
+  const netId = NETWORK[net]
+  if (!netId && netId !== 0) throw new Error('Not a valid network name')
+  return netId
+}
+
+export class EventID {
+  protected readonly _tag = TAG
+
+  private readonly _networkID: number
+  private readonly _separator: Uint8Array
+  private readonly _controller: Uint8Array
+  private readonly _init: Uint8Array
+  private readonly _eventHeight: number
+  private readonly _event: CID
+
+  // WORKARDOUND Weird replacement for Symbol.hasInstance due to
+  // this old bug in Babel https://github.com/babel/babel/issues/4452
+  // which is used by CRA, which is widely popular.
+  static isInstance(instance: any): instance is EventID {
+    return typeof instance === 'object' && '_tag' in instance && instance._tag === TAG
+  }
+
+  /**
+   * Create a new EventID.
+   */
+  constructor(networkID: number, separator: Uint8Array, controller: Uint8Array, init: Uint8Array, eventHeight: number, event: CID) {
+    this._networkID = networkID
+    this._separator = separator
+    this._controller = controller
+    this._init = init
+    this._eventHeight = eventHeight
+    this._event = event
+  }
+
+  /**
+   * Create a new EventID.
+   */
+  static create(networkID: number | string, separator: string, controller: string, init: CID | string, eventHeight: number, event: CID | string ): EventID {
```

Comment:
Can we allow `separator` to be a bytestring as well? In the case of a model, we will get it as a Uint8Array when we read it from the initevent.

cc @AaronGoldman thoughts about hashing either a string or a bytearray here?

Reply:
The event is dag-cbor map with a field in it. I think we should take whatever is there. If it is a cbor Text string hash the utf-8 bytes, if it is a  Byte string hash the Uint8Array.